library/alloc/src/vec.rs

```diff
@@ -2391,11 +2384,24 @@ where
 }
 
 impl<T> SpecExtend<T, IntoIter<T>> for Vec<T> {
-    fn spec_extend(&mut self, mut iterator: IntoIter<T>) {
-        unsafe {
-            self.append_elements(iterator.as_slice() as _);
+    fn spec_extend(&mut self, iterator: IntoIter<T>) {
+        // Avoid reallocation if we can use iterator's storage instead. This requires 1 memcpy and 0-1 memmove
+        // while reallocation would require 1 alloc, 1-2 memcpy, 1-2 free
+        if mem::size_of::<T>() > 0
+            && self.capacity() - self.len() < iterator.len()
+            && iterator.cap - iterator.len() >= self.len()
+        {
+            // Safety: we just checked that IntoIter has sufficient capacity to prepend our elements.
+            // Prepending will then fill the uninitialized prefix.
+            *self = unsafe {
+                let mut v = iterator.into_vec_with_uninit_prefix(self.len() as isize);
+                ptr::copy_nonoverlapping(self.as_ptr(), v.as_mut_ptr(), self.len);
+                self.set_len(0);
```

Comment:
I think this would be clearer if we lifted it out of this block. It's trickier to reason about an assignment to `self` in a block that also uses `self`.

Reply:
To clarify, I mean replacing `*self = unsafe { .. }` with `let v = unsafe { .. }; *self = v;`

Reply:
done